### PR TITLE
DBZ-2495 Updating to handle captured_column_list without specifying column.exclude.list

### PR DESCRIPTION
**Pull request by jgormley6**
https://issues.redhat.com/browse/DBZ-2495

Additional Changes:
Fixed the SqlServerSnapshotChangeEventSource#checkExcludedColumns to actually filter Previously it was just adding all the columns regardless and not utilizing the filters properly.

Fixed Testing.Debug#disable to flip the flag to false.

Removed `private Table sourceTable;` from SqlServerChangeTable as it was hiding the underlying sourceTable in `ChangeTable`.

### CODE REVIEW

**Comment by jgormley6** [comment]
Rename test is a little flakey. It passes locally. Looking into it

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTable.java
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTable.java:39]: Can be `final`? Also simply `capturedColumns` would be preferred, no need to encode the type into the name (same for the getter exposing it).

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTablePointer.java
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTablePointer.java:104]: This will make the map not big enough due to the default load factor, triggering a re-size if all columns are added. I wouldn't bother and just call new `HashMap<>()`.

File: debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java:8]: Could you also update `SqlServerConnectorIT#excludeColumnWhenCaptureInstanceExcludesColumns()` and remove the filter? As discussed in the Jira issue, this should not be needed any longer with your fix.

**Review by gunnarmorling — COMMENTED** [review]
@jgormley6, thanks a lot; some first cosmetic comments inline.

Taking a step back though, could you provide some more high-level explanation of the change? It seems to be quite involved and I'd really like to fully understand this one. My key question is: is it really wise to have the `Table` metadata not contain all table columns? Either for SQL Server or for Oracle (I'm not sure which one) we already adjust the snapshot SELECT statement to only fetch the actually included columns. I.e. there's code in place which handles a result set which hasn't the full span of the table. I'm wondering what problem you encountered which made this unfeasible to do the same. AFAIK, we can obtain the column name from result set columns, so we should be able to populate the full-width row also based on non-full-width result set? Thanks a lot for elaborating!

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTable.java
Comment (jgormley6) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTable.java:39]: Yeah, I can make that change. I had both the original string and a list at one point but with the string gone, you're right the "list" is unnecessary.

**Comment by jgormley6** [comment]
@gunnarmorling 
I've got a few posts on the JIRA explaining the alternatives I explored.
https://issues.redhat.com/browse/DBZ-2495?focusedCommentId=15130079&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-15130079

The main reason for editing the schema is the TableSchemaBuilder. I found it very difficult to create hooks into it to additionally filter extra columns based on the CDC tables without a major rework of the code/effecting all of the other relational connectors in a significant way. 
The two problems are the Object[] being passed to the generated functions and the schema for the event being generated in the create method.
I needed to somehow filter out columns from the schema, but passing in the additional parameters for the CDC table means I need to make changes for all relational connectors.
The Object[] that gets passed lacks information and makes things difficult for the code to understand if something is "null" or if something should not be included in the result without passing additional parameters.

This prevents us from doing the following without passing more information, otherwise we will take a "null" and default the value rather than exclude the column entirely.

> we should be able to populate the full-width row also based on non-full-width result set


I think my most preferred way of handling this would be to somehow wrap the `columnNameFilter`, but that also turned out to not be simple to do on a per table basis.

In regards to:

> Either for SQL Server or for Oracle (I'm not sure which one) we already adjust the snapshot SELECT statement to only fetch the actually included columns

If you mean the `SqlServerSnapshotChangeEventSource#checkExcludedColumns`, the original implementation did not actually filter anything and would fail in certain cases if it did. (due to the same reason I fixed in  https://issues.redhat.com/browse/DBZ-2522). I'm not sure about Oracle, but it would likely face the same issue without the changes made to `RelationalSnapshotChangeEventSource` for the data to be positioned based instead of raw index based.

To explain my changes more directly, we need to filter out columns not included in the CDC tables in two places. When we first generate all the schemas in `SqlServerSnapshotChangeEventSource#readTableStructure`, and whenever a schema is updated in `SqlServerConnection.#getTableSchemaFromTable`. This allows for schemas to be updated "hot" or "cold". 

I felt that it was fine to not include all the columns in the Table metadata as we really only care about the CDC table structure. We just happen to also need the source table to get all the information that we need. 

I fixed the filtering in `SqlServerSnapshotChangeEventSource#checkExcludedColumns` to allow for the snapshot to also exclude those columns.
I needed to create `TableEditorSkippableColumnsImpl` to allow for columns to be excluded as the original implementation updated the column positions.

I kind of rambled this out (sorry!), let me know if I should explain things more clearly anywhere!

**Comment by gunnarmorling** [comment]
Thanks for elaborating, that helps, @jgormley6. I understand the problem now, the Object[] with values received during streaming won't tell you which columns it contains.

I'm still unsure though whether the shape of the capture instance should determine the shape (number of fields) of Kafka message, or rather the source table. I'm inclined towards the latter, as it'd seem confusing to impact the shape of messages produced from snapshotting (scanning the source table) based on the shape of the capture table, wouldn't it? In other words, I feel the shape of the source table should be the deciding factor, and in case of a narrower capture instance, any missing columns should just be null/defaulted in the produced change events.

Now that's a consideration in terms of semantics/design, I see your point about implementing it, though. Would it be a possibility for the change event emitter to produce Object[] of the full table width, expanding a narrower array obtained from the capture table with null values for any missing columns, based on the current `TableSchema` we have? This might be a much more local change than other options potentially. WDYT?

**Comment by jgormley6** [comment]
Hmm, personally I think the capture tables should dictate what's captured as that's what is being specified as to be "captured". 

I can't really think of a case where someone would want to have data as part of the snapshot, but not as part of the capture changes. They would get "read" events saying something existed, but then any subsequent "updated" events would all have null/defaults. 

My idea of this was to act almost exactly like using exclude/include columns, but being defined by the database instead of part of the connector configuration. Without removing the un-captured columns from the schema, we would still need to specify the exclude/include column and so this PR wouldn't really be helping us much beyond minimizing the data in the snapshot select.

For expanding the Object[], that's sort of what's currently happening. If you have a table of [colA, colB, colC, colD], and exclude colB and colD, the object array will be ["colAData", null, "colCData"]. It's just only going up to the last needed position for the expansion.

**Comment by gunnarmorling** [comment]
> I think the capture tables should dictate what's part of the schema as that's what is being specified as to be "captured".

Interesting, so this actually is growing on me. In this light, wouldn't the best way to address the issue then to drive metadata retrieval based on the capture table to begin with? Then there wouldn't be any mismatch between the schema we have and the span of results we get by definition. We'd only have to reach out to the original table to obtain nullability info for the columns, I believe. WDYT? // CC @jpechane, your toughts welcome, too!

**Comment by jgormley6** [comment]
That sounds best to me too. I'll start looking into making that change.

**Comment by jgormley6** [comment]
I updated the Tables class to utilize TableEditor when creating a new Table. This is why the positions were off when creating the Object[] and when accessing them in the TableSchemaBuilder. Those should match now and give us the same functionality as building the schema directly from the CDC tables with a lot less hassle.
This allows me to remove that skippable thing I built as well. 

I can remove what I did with the column position indexing and go back to raw indexing now, but I think using position still is probably a safer approach, but I can revert that if desired.

**Comment by jpechane** [comment]
@jgormley6 @gunnarmorling One question - I probably lost the reason why it is not wise but can't we solve the issue just by having the connector to add the columns missing from capture tables to columns blacklist withou touching any other code?

**Comment by jpechane** [comment]
Regarding the proposal on getting the metadata from capture table instead of source table. Right now the connector supports taking the snapshot of source tables that don't have capture tables -those are then omitted from streaming and IIRC the streaming start as soon as capture tables are created. Would this functionlity be still supported then?

**Comment by jgormley6** [comment]
The column filtering is currently done through a more global variable and not on a per table basis. I looked into updating it to be on a per table basis, but I found it difficult to create hooks to allow that without making significant changes to method signatures for the relational classes in debezium-core. 
The main issue around it was that we create the schemas/filter, pass them around to various places (like TableSchemaBuilder), and then later find out what's the actual structure of the CDC tables during the polling phase. So I would need to somehow pass that information through.
CDC being updatable on the fly also presented a challenge.

**Comment by jgormley6** [comment]
I think the update I just made helps with what @gunnarmorling was wanting, but still allows for snapshots of tables without cdc enabled.
With the TableEditor being used, the columns are shifted. The schema will now match the CDC table (or the original table if there is no CDC table).
I also think this is preferable to directly interacting with the CDC columns as it means we don't have to loop through all the columns twice. If we did it based on the CDC tables, we would still need to also loop through the source table columns to get the nullability.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTable.java
Comment (jpechane) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTable.java:49]: IMHO this should not be done in constructor but on the caller side. The string encoding is the stored procedure implementation detail and should not leak into logical representation of the capture table.

Comment (jgormley6) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTable.java]: Agreed, updated it

File: debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java:321]: Can we still keep it in a separate method?

File: debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java:324]: This is very similar to the code in https://github.com/debezium/debezium/pull/1822/files#diff-d9a5092f0eb1f68e8881b17b09f25827R106

Can't it be extracted into a utility method shared?

File: debezium-core/src/main/java/io/debezium/relational/TableImpl.java
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/relational/TableImpl.java:100]: This is just a technical method, I don't think this belongs to the interface. Is not this unnecessary code optimization?

**Comment by jpechane** [comment]
@jgormley6 Nice work, I just left a couple of comments related to minor issues
@gunnarmorling I think we are good to go

File: debezium-core/src/main/java/io/debezium/relational/TableImpl.java
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/TableImpl.java:100]: Indeed, let's not expose it on the interface. `toString()` shouldn't be invoked at runtime anyways, so shouldn't make any difference.

File: debezium-core/src/main/java/io/debezium/relational/Tables.java
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/Tables.java:185]: Nice, this was bothering me, too :)

File: debezium-core/src/main/java/io/debezium/relational/TableImpl.java
Comment (jgormley6) [reply debezium-core/src/main/java/io/debezium/relational/TableImpl.java:100]: This was to allow `Tables` to utilize `Table` over `TableImpl` (it utilizes the method for it's `toString())`. I can rework the method to utilize the normal toString or throw in casting instead.

File: debezium-core/src/main/java/io/debezium/relational/Tables.java
Comment (jgormley6) [reply debezium-core/src/main/java/io/debezium/relational/Tables.java:185]: It actually causes a bug with the rest of the implementation because the TableEditor shifts the positions of the columns (something I found unexpected).

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/Tables.java:185]: Interesting; so are you reverting this one then?

What's unclear to me is why we need that explicit position field in `Column` in the first place; I'd assume we'd always be able to derive that info from the list index. But seems rather risky to change right now and it's definitely a separate discussion.

File: debezium-core/src/main/java/io/debezium/relational/TableImpl.java
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/TableImpl.java:100]: The latter sounds good.

File: debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java:326]: So in which case would `greatestColumnPosition` be different from the width of the result set (ignoring any +/-1 difference due to indexing based on 0 vs. 1)?

File: debezium-core/src/main/java/io/debezium/relational/Tables.java
Comment (jgormley6) [reply debezium-core/src/main/java/io/debezium/relational/Tables.java:185]: Sorry, I meant that I needed to make that change to solve the bug. 

I prefer using the position as it seems to prevent issues around excluding columns (that's what my past three PRs have been about 😅 ). It just feels a little safer to me than the raw index.

File: debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java
Comment (jgormley6) [reply debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java]: They should be the same now, I mentioned in a previous comment:

> I can remove what I did with the column position indexing and go back to raw indexing now, but I think using position still is probably a safer approach, but I can revert that if desired.

Comment (jgormley6) [reply debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java]: I feel position is a safer way to go, as I've been implementing several changes around this area, and each time I run into issues around these indexes. Using the position everywhere ensures we have a consistent column size and placement regardless of what happens (columns being excluded and what not).

Comment (jgormley6) [reply debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java]: Actually, this might still be necessary now that I think about it.
With the snapshot now filtering items out of the select based on the include/exclude list, if you filter out a column in the middle of the row, it'll run into issues with the raw index...

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java]: Ok, I suppose you're right. Overall I still have the feeling (or hope?) we could get rid of the position business altogether, but that'll have to wait for another time. This first loop is invoked only once per result set, so overhead is small.

**Comment by jgormley6** [comment]
Only test that failed in the last run was a postgres test that has previously passed and I don't think is related to these changes. Let me know if there's any additional clean up I can do before merging.

**Comment by gunnarmorling** [comment]
@jgormley6, yes, sorry about the noise with flaky tests, it's an ongoing struggle we're currently going through.

On this PR itself, it's good to go, I'll do some polishing and this will then be part of tomorrow's 1.3.0.Final release. Note I'm still somewhat doubtful about the need for all the column position/index mapping; is e.g. the `IndicesMapping` needed now, as the table metadata is matching the CDC table structure. I'll log a separate issue for getting to the grounds of this, hoping we might be able to simplify things a bit. But the test coverage is good, so I'm confident enough it's sound on the functional level.

**Comment by gunnarmorling** [comment]
Logged https://issues.redhat.com/browse/DBZ-2612 as a follow-up to verify the column position indirection is truly needed. @jgormley6, your input on this by commenting on that issue would be very welcomed. Thanks!

Pushed one more commit; will merge after CI has hopefully passed.

**Review by gunnarmorling — APPROVED** [review]

**Comment by gunnarmorling** [comment]
CI green, for once :) Applying; thanks, @jgormley6!

**Pull request merged**
